rfcs/0020-project_api_extend_metadata.md

```diff
@@ -0,0 +1,252 @@
+
+
+## On "required" and "optional" metadata
+
+Currently even though all options available for a given project can be
+discovered via the Project API 'server_info_query' interface there is no way to
+known which options belong (or apply) to which API method (like the
+generate_project, build, flash, and open_transport methods).
+
+This is fine when the user knowns beforehand which method accepts a set of
+options, so it's possible to manually select which options will be passed to a
+given API method, like when using the API in a Python script.
+
+However that's a problem when the API user (e.g. TVMC) needs to automatically
+determine the options available for the API methods, like when automatically
+building a command line parser with subcommand domains that closely mapped to
+the API methods (e.g. subcommands to create, build, and flash a project).
+
+Moreover, currently it's impossible to determine which option is required and
+which one is optional, so it would be at least necessary for the API user to
+build a static ad hoc table with all options available on a given project
+stating which option is required and which one is optional for the project. This
+is impractical to maintain and would result in the API user having to update the
+the static table every time an option is added, removed, or modified in the
+Project API server.
+
+Hence to ease the automatic detection of the options available on each Project
+API method the following two new metadata are proposed: 'required' and 'optional'.
+
+Both will contain a list of method names for which the option is available,
+either as a required option (if in 'required' list), or as an optional option
+(if in 'optional' list). At least one API method must be listed in 'required' or
+in the 'optional list. A method name must be listed only in the 'required' or in
+the 'optional' list, i.e. an option can't be require and optional at the same
+time for given API method. An option can be required for a method and optional
+for another method.
+
+The elements in the lists 'required' and 'optional' must be in the set of method
+names implemented by the ProjectAPIClient class and that have the parameter
+'options' defined. These methods are dispatched to the server, which implements
+the server counterparts to properly handle the client dispatches and
+ultimately defines the options available for each API method. The current method
+name that satisfy these criteria are 'generate_project', 'build', 'flash', and
+'open_transport'.
+
+'required' metadatum or 'optional' metadatum (or both) must be provided for
+every option.
+
+
+## On "type" metadatum
+
+The option type can sometimes be determined implicitly by what's return
+in metadatum 'choices', but this not ideal. For example, for option 'verbose' it
+would be possible to infer it is a boolean option and therefore it can be
+converted to a command line flag if metadatum 'choices' is a couple of True and
```

Comment:
```suggestion
in metadatum `choices`, but this not ideal. For example, for option `verbose` it
would be possible to infer it is a boolean option and therefore it can be
converted to a command line flag if metadatum `choices` is a couple of True and
```